Integrated HuggingFace inference API for summarization and Perspective Generation.

ParagGhatage opened the pull request:
Fixes: #27 

## Tasks completed :
 - Connected frontend and backend.
 - integrated HuggingFace Inference API for summarization and generating opposite perspective.
 - integrated scraper and summarization functions in FastAPI routes to handle requests.
 
-  setup summarization_service and perspective-service to make api calls with prompt.

- Updated Readme with additional instructions to setup .env .

- tested entire workflow from entering URL to generating Summary and opposite perspective.

- Integrated small prompt in this branch only to ensure smooth flow.

## Models Used:

I tested multiple HuggingFace models for summarization and perspective generation tasks and chose best models for our tasks.

- Summarization: `google/bigbird-pegasus-large-arxiv`
- Opposite perspective generation: ` mistralai/Mixtral-8x7B-Instruct-v0.1`

## Demo video:


https://github.com/user-attachments/assets/e6ed7262-20fa-4cd8-98ee-518434a83fb8


## Collaborators:
thanks @Ashutosh-KARNX7  for collaboration on Model Selection.

ParagGhatage commented:
Hey @ManavSarkar ,

Model that I have integrated for summarization doesn't need a prompt, it just takes text and summarizes it.
So, we dont need summarization prompt for now.

I have updated Perspective genration prompt in #28  to fit these models and workflow.

ParagGhatage commented:
@ManavSarkar , resolved the merge conflicts.

PranshuLakhotia commented:
@ParagGhatage @ManavSarkar  I see that you have chosen two models for summarizing and opposite pers generation. We can work with only one model for reducing the API calls(increasing latency) or we could create a LangGraph architecture where we can use it to create a summary agent whose output is given to the input of perspective agent.
Also I had raised a PR for the perspective generation part using Mistral so you can raise this PR on top of my PR.

PranshuLakhotia reviewed (changes requested):
Also I reviewed your code and I didn't find connection between the summary generated and the opposing perspective. Opposing perspective should be with reference to the summary generated. So I think that summary should be sent via API call when the perspective is being generated.

ParagGhatage commented:
> @ParagGhatage @ManavSarkar I see that you have chosen two models for summarizing and opposite pers generation. We can work with only one model for reducing the API calls(increasing latency) or we could create a LangGraph architecture where we can use it to create a summary agent whose output is given to the input of perspective agent. Also I had raised a PR for the perspective generation part using Mistral so you can raise this PR on top of my PR. I had worked on manav's suggestion about the regex.

@PranshuLakhotia ,
There are 3 things that need to be cleared-

1. HuggingFace inference API counts requests per Secret Token.
So, using 2 requests to 2 seperate models or 2 requests to same models is the same.

2. @ManavSarkar , have alredy mentioned not to implement Langchain andLangGraph for now.

3. My PR making the entire website Up and running , so, make your changes in model after merger of this PR.It will be better.

ParagGhatage commented:
> Also I reviewed your code and I didn't find connection between the summary generated and the opposing perspective. Opposing perspective should be with reference to the summary generated. So I think that summary should be sent via API call when the perspective is being generated.

@PranshuLakhotia , There is a connection.

in `frontend/src/app/article/page.tsx`

1. first request is made to `scrape-clean-summarize` endpoint with url which returns summary of the article.
2. this   text is saved in state variable named summaryText and second request is made to `generate-perspective ` endpoint with summary as a body parameter, which returns opposite perspective.

PranshuLakhotia commented:
@ParagGhatage  I understand your concern. The current approach calls two separate models, one to generate a summary and another to produce an opposite perspective. 
When you summarize an article first you lose a lot of context that might be important for generating a truly balanced or opposite perspective. This two model approach can lead to issues such as:

Context Loss: The summary may not capture all nuances resulting in a perspective that doesn't fully address the original article's arguments.
The perspective generation model may work on an incomplete picture, leading to poorer quality outputs compared to if it had access to the full text or a more context-rich input.
Errors or biases introduced during summarization can compound when used as input for the perspective generation further skewing the final output.

A more effective solution could be to provide the full article context to the second model or even design a single model that returns both the summary and an opposing perspective in a single, coherent response. This would likely enhance the overall result while reducing the number of API calls.

ParagGhatage commented:
> @ParagGhatage I understand your concern. The current approach calls two separate models, one to generate a summary and another to produce an opposite perspective. When you summarize an article first you lose a lot of context that might be important for generating a truly balanced or opposite perspective. This two model approach can lead to issues such as:
> 
> Context Loss: The summary may not capture all nuances resulting in a perspective that doesn't fully address the original article's arguments. The perspective generation model may work on an incomplete picture, leading to poorer quality outputs compared to if it had access to the full text or a more context-rich input. Errors or biases introduced during summarization can compound when used as input for the perspective generation further skewing the final output.
> 
> A more effective solution could be to provide the full article context to the second model or even design a single model that returns both the summary and an opposing perspective in a single, coherent response. This would likely enhance the overall result while reducing the number of API calls.

Let's see what @ManavSarkar thinks about this.
if he says,  i will update the code to send entire article text to perspective generation instead of summary.

PranshuLakhotia commented:
I had already implemented the feature in my local as this issue was assigned to me. This leads to doubling of work and effort. I believe we should collaborate rather than raising PR on issues assigned to others. Lets keep the spirit of Open Source maintained.

PranshuLakhotia commented:
> > Also I reviewed your code and I didn't find connection between the summary generated and the opposing perspective. Opposing perspective should be with reference to the summary generated. So I think that summary should be sent via API call when the perspective is being generated.
> 
> @PranshuLakhotia , There is a connection.
> 
> in `frontend/src/app/article/page.tsx`
> 
> 1. first request is made to `scrape-clean-summarize` endpoint with url which returns summary of the article.
> 2. this   text is saved in state variable named summaryText and second request is made to `generate-perspective ` endpoint with summary as a body parameter, which returns opposite perspective based on article summary.

Yes saw that

amanverma-92 commented:
@ParagGhatage, @makhanovable. I analyzed your code and I saw that one of the models you are using is facebook/bart-large-cnn. It is trained on 406M params which is rather insufficient for a better result. Instead we could use mistral works or Phi3mini which would give a better result.

ParagGhatage commented:
@PranshuLakhotia , @amanverma-92 ,
Thanks for your suggestion regarding model.

I have changed code to use `mistralai/Mixtral-8x7B-Instruct-v0.1` for perspective generation.
and also added small concise prompt to to maintain the flow.

@PranshuLakhotia , You can go ahead and replace that prompt with prompt from #28 after this PR gets merged.

sagarkawad commented:
Running the backend and passing the url in the body using a post request to the /scrape-and-summarize endpoint, gives me this error. 
<img width="1466" alt="Screenshot 2025-03-03 at 7 19 00 PM" src="https://github.com/user-attachments/assets/4269e47f-c472-41fd-9755-891ae8e633ac" />

sagarkawad commented:
> Running the backend and passing the url in the body using a post request to the /scrape-and-summarize endpoint, gives me this error. <img alt="Screenshot 2025-03-03 at 7 19 00 PM" width="1466" src="https://private-user-images.githubusercontent.com/73158739/418600578-4269e47f-c472-41fd-9755-891ae8e633ac.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEwMTA1MjQsIm5iZiI6MTc0MTAxMDIyNCwicGF0aCI6Ii83MzE1ODczOS80MTg2MDA1NzgtNDI2OWU0N2YtYzQ3Mi00MWZkLTk3NTUtODkxYWU4ZTYzM2FjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzAzVDEzNTcwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYwNTY0OTU0ODE3YzI0NzlmYWMxOGEzODJlNjFkYzM3ODhhZGUwM2RiYTE5YWMxZWE1MjMxNDAyMzVjMDFlZjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.eOF_B7yugODbPP-UkfvH4W95YLGSelSzTiisE7tIDvc">

would like to work on this if the PR gets merged.

ParagGhatage commented:
> Running the backend and passing the url in the body using a post request to the /scrape-and-summarize endpoint, gives me this error. <img alt="Screenshot 2025-03-03 at 7 19 00 PM" width="1466" src="https://private-user-images.githubusercontent.com/73158739/418600578-4269e47f-c472-41fd-9755-891ae8e633ac.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEwMTE2MTQsIm5iZiI6MTc0MTAxMTMxNCwicGF0aCI6Ii83MzE1ODczOS80MTg2MDA1NzgtNDI2OWU0N2YtYzQ3Mi00MWZkLTk3NTUtODkxYWU4ZTYzM2FjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzAzVDE0MTUxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWViMmMyMjIwNmFlYjAxOGFiZTgxODYxMTRhZjE1MDA2MzM2YmZmODVhOGJlOTgyYzE1Zjk3YWNiMTEwN2I2ZmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.IwwbIVqFociaGH8Kqu1WFx1tBBkzQUM9DrZ36hdqrkg">

It could be related to the summarization .
I will look into it.

Thanks for pointing that out.

ParagGhatage commented:
@sagarkawad , the issue was in summarization model, which was not able to handle large articles,

But, I have changed the model.
Now, everything works great.

ParagGhatage commented:
@ManavSarkar ,  I have added model `google/bigbird-pegasus-large-arxiv` for summarization, which can handle any amount of text from large article to research paper.

It's working great.

ManavSarkar commented:
@PranshuLakhotia you have pointed it right, but let's merge this since already the conversation is going long.
Also, lots of files are changed. Let's keep it for another issue.
